test-common/src/main/java/org/apache/kafka/common/test/TestUtils.java

```diff
@@ -125,4 +148,131 @@ static void retryOnExceptionWithTimeout(final Runnable runnable) throws Interrup
+    }
+
+    public static MemoryRecords singletonRecords(byte[] value, byte[] key) {
+        return singletonRecords(value, key, Compression.NONE, RecordBatch.NO_TIMESTAMP, RecordBatch.CURRENT_MAGIC_VALUE);
+    }
+
+    public static MemoryRecords records(List<SimpleRecord> records,
+                                        byte magicValue,
+                                        Compression codec,
+                                        long producerId,
+                                        short producerEpoch,
+                                        int sequence,
+                                        long baseOffset,
+                                        int partitionLeaderEpoch) {
+        int sizeInBytes = DefaultRecordBatch.sizeInBytes(records);
+        ByteBuffer buffer = ByteBuffer.allocate(sizeInBytes);
+
+        try (MemoryRecordsBuilder builder = MemoryRecords.builder(
+            buffer,
+            magicValue,
+            codec,
+            TimestampType.CREATE_TIME,
+            baseOffset,
+            System.currentTimeMillis(),
+            producerId,
+            producerEpoch,
+            sequence,
+            false,
+            partitionLeaderEpoch
+        )) {
+            records.forEach(builder::append);
+            return builder.build();
+        }
+    }
+
+    public static MemoryRecords records(List<SimpleRecord> records, byte magicValue, Compression codec) {
+        return records(records,
+            magicValue,
+            codec,
+            RecordBatch.NO_PRODUCER_ID,
+            RecordBatch.NO_PRODUCER_EPOCH,
+            RecordBatch.NO_SEQUENCE,
+            0L,
+            RecordBatch.NO_PARTITION_LEADER_EPOCH);
+    }
+
+    public static int waitUntilLeaderIsElectedOrChangedWithAdmin(Admin admin,
+                                                                 String topic,
+                                                                 int partitionNumber,
+                                                                 long timeoutMs) throws Exception {
+        GetPartitionLeader getPartitionLeader = (t, p) -> Optional.ofNullable(getLeaderFromAdmin(admin, t, p));
+        return doWaitUntilLeaderIsElectedOrChanged(getPartitionLeader, topic, partitionNumber, timeoutMs);
+    }
+
+    private static Integer getLeaderFromAdmin(Admin admin, String topic, int partition) throws Exception {
+        TopicDescription topicDescription = admin.describeTopics(Collections.singletonList(topic)).allTopicNames().get().get(topic);
+        return topicDescription.partitions().stream()
+            .filter(partitionInfo -> partitionInfo.partition() == partition)
+            .findFirst()
+            .map(partitionInfo -> partitionInfo.leader().id() == Node.noNode().id() ? null : partitionInfo.leader().id())
+            .orElse(null);
+    }
+
+    private static int doWaitUntilLeaderIsElectedOrChanged(GetPartitionLeader getPartitionLeader,
```

Comment:
Could you please replace `GetPartitionLeader` by `BiFunction<String, Integer, Optional<Integer>>`?